src/coreclr/vm/threadsuspend.cpp

```diff
@@ -3004,27 +3011,50 @@ BOOL Thread::RedirectCurrentThreadAtHandledJITCase(PFN_REDIRECTTARGET pTgt, CONT
     // Check to see if we've already got memory allocated for this purpose.
     CONTEXT *pCtx = GetSavedRedirectContext();
 
-    // If we've never allocated a context for this thread, do so now
+    // If we've never assigned a context for this thread, do so now
     if (!pCtx)
     {
-        pCtx = new (nothrow) CONTEXT();
-
-        if (!pCtx)
-            return (FALSE);
-
-        // Save the pointer for the redirect function
-        _ASSERTE(GetSavedRedirectContext() == NULL);
-        SetSavedRedirectContext(pCtx);
+        pCtx = m_pSavedRedirectContext = AllocateOSContextHelper(&m_pOSContextBuffer);
+        _ASSERTE(GetSavedRedirectContext() != NULL);
     }
 
     //////////////////////////////////////
     // Get and save the thread's context
 
+#if defined(TARGET_X86) || defined(TARGET_AMD64)
+    _ASSERTE((pCtx->ContextFlags & CONTEXT_XSTATE) ==
```

Comment:
Cannot we use the CopyContext API?

Reply:
I do not know if it would be always safe to use `CopyContext`. `CopyContext` doc mentions that the context must be initialized with `InitializeContext`. We fall back to regular `CONTEXT` when we do not have AVX.

Reply:
Actually I noticed that we still use `InitializeContext` even if AVX is disabled and it does not cost us much extra. Therefore I will use `CopyContext`. It will make this part a simpler.